libs/platform/src/lib/components/form/form-generator/form-generator.service.ts

```diff
@@ -0,0 +1,402 @@
+})
+export class FormGeneratorService {
+
+    /**
+     * @hidden
+     * @private
+     */
+    private _formComponentDefinitions: FormComponentDefinition[] = DEFAULT_COMPONENTS_LIST;
+
+    /**
+     * @hidden
+     * @private
+     */
+    private _validationErrorHints = DEFAULT_VALIDATION_ERRORS;
+
+    /**
+     * @hidden
+     * @private
+     */
+    private _latestAsyncValue: any;
+
+    constructor(
+        private _fb: FormBuilder
+    ) {
+    }
+
+    /**
+     * @description Generates `FormGroup` class with the list of `DynamicFormControl` control classes
+     * defined in `formItems` argument.
+     * @param formItems the list of form items which used for form control generation.
+     * @returns `FormGroup` class with the list of `DynamicFormControl` control classes.
+     */
+    async generateForm(formItems: DynamicFormItem[]): Promise<FormGroup> {
+        const form = this._fb.group({});
+
+        formItems.forEach((q) => {
+
+            const formItemComponentType = this.getComponentDefinitionByType(q.type);
+
+            if (!formItemComponentType) {
+                console.warn(`Form item '${q.name}' with type '${q.type}' has no defined component.
+                Please use 'addComponent' method in ${FormGeneratorService.name} to define appropriate relations.`);
+
+                return;
+            }
+
+            let validator: AsyncValidatorFn = null;
+
+            if (isFunction(q.validate)) {
+                validator = async (control: DynamicFormControl) => {
+                    const formValue = await this.getFormValue(form);
+
+                    const obj = q.validate(control.value, formValue);
+
+                    const result = await this._getFunctionValue(obj);
+
+                    return result === true ? null : { validator: typeof result === 'boolean' ? true : result };
+                };
+
+                q.asyncValidators = q.asyncValidators || [];
+                q.asyncValidators = typeof q.asyncValidators === 'function' ? [q.asyncValidators] : q.asyncValidators;
+
+                q.asyncValidators.push(validator);
+            }
+
+            let formControl: DynamicFormControl;
+
+            q.validators = q.validators || [Validators.nullValidator];
+
+            formControl = new DynamicFormControl(q.default, {
+                validators: q.validators,
+                asyncValidators: q.asyncValidators,
+                dynamicFormitem: q,
+                updateOn: 'change'
+            });
+
+            form.addControl(q.name, formControl);
+
+            formControl.updateValueAndValidity({ emitEvent: false });
+        });
+
+        const formControlNames = Object.keys(form.controls);
+
+        // After we created initial form, we need to get dynamic labels, choices and value
+        for (const key of formControlNames) {
+            const formItem = formItems.find(fi => fi.name === key);
+
+            formItem.message = await this._getFormItemLabel(formItem, form);
```

Comment:
are we overwriting the `message` option here? Looks quite unsafe

Reply:
But in this case we overwrite user's config directly.
That means if the user will try to use this config second time he will get a different result

Reply:
Changed the behaviour. Now we clone input array so we don't mess with original passed value